DESIGN.md

```diff
@@ -266,6 +278,18 @@ This command checks to see if there are new versions of `swiftly` itself and upg
 
 `swiftly self-update`
 
+### Proxy command invocations to a requested toolchain
+
+Swiftly will create a set of symbolic links in its SWIFTLY_BIN_DIR during installation that point to the swiftly binary itself for each of the common toolchain commands, such as swift, swiftc, clang, etc. This mechanism will allows swiftly to proxy those command invocations to a requested toolchain. A toolchain can be requested in these ways in order of precedence:
+
+* Special toolchain selectors among the regular tool command-line arguments (e.g. `swift build +5.10.1`) with the special '+' prefix
+* The presence of a .swift-version file in the current working directory, or ancestor directory, with the required toolchain version
+* The swiftly default (in-use) toolchain set in the config.json by `swiftly install` or `swiftly use` commands
+
+In the first two cases, if there is no matching toolchain installed, swiftly will attempt to automatically install the requested toolchain and use it if the installation succeeeds.
+
+Note: If swiftly automatically installs a toolchain during proxying and that toolchain requires post installation steps then the proxy will abort with those post installation instructions.
```

Comment:
Note: IDEs like VSCode need to know the actual path to the swift executable (currently VSCode follows the symbolic links) so it can use other components of the selected toolchain. We might need to add a new subcommand to provide this information.

Reply:
Ah, so this is why my vscode experiments with swiftly have failed so far. Can you provide more details of what sorts of details (components) the vscode needs from the toolchain? Maybe we can abstract that like we do the location of the toolchain executables in swiftly.

Reply:
All we need is the path to the toolchain. It then uses this to find various executables (lldb-dap, sourcekit-lsp, llvm-cov) which I can see being abstracted, we also need the path to lldb dynamic library to pass to CodeLLDB. Not sure how you'd do that. There might be other things I cannot remember offhand.

If we add a subcommand to provide this information, we would also need to edit the extension of course to use this information.

Reply:
Thanks, I can try and dig deeper about the lldb dylib issue.

It seems to me that swiftly is basically an abstract toolchain, and should be able to cleverly pretend that it is a static toolchain in most (maybe all) usages. We could provide a subcommand to give use the real toolchain location, but it feels like an escape hatch from the abstraction. I'd like to understand why swiftly couldn't provide the complete abstraction.

Reply:
Ok, I see now that there are cases where swiftly's toolchain abstraction isn't going to be sufficient. I've elaborated on the boundaries of swiftly's abstraction: just the toolchain binaries, and not for long-lived invocations. There's an escape hatch through a special `swiftly use --location` flag that will provide the physical location of a toolchain, and also help for changes in toolchain selection. It does mean that if tools operate outside the boundaries then they might need to be aware of swiftly and interface with it to provide the best user experience.

Reply:
I'm not looking to override the location of the toolchain. I'm looking for the location of the  current toolchain.

Reply:
If you invoke `swiftly use` without a selector today it will tell you the version number of the toolchain that is in use in the current context. I'm proposing the addition of the `--location` flag here to make that output the file path of the toolchain that is in use. It probably wouldn't make sense for the flag to be used when you provide a selector, so that can be checked in the subcommand and documented.